### PR TITLE
Port :string.replace/3 and :string.replace/4 to JS

### LINKED ISSUE #424: Port :string.replace/4 to JS

### PR DESCRIPTION
**Pull request by Sancxo**
Close #424 

## Summary by CodeRabbit

* **New Features**
  * Added string replacement utility that returns segmented lists with modes: all, leading, trailing; validates inputs, raises clear errors, and preserves the original when pattern is empty or absent.

* **Tests**
  * Added extensive JS and Elixir test suites covering normal behavior, directions, Unicode, consecutive patterns, edge cases, and error scenarios.

* **Chores**
  * Updated internal call mappings to integrate the new replacement behavior.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by bartblast** [comment]
@Sancxo Could you fix the CI issue: 

```
Run mix eslint

/home/runner/work/hologram/hologram/test/javascript/erlang/erlang_test.mjs
  41:7  error  'integer36' is assigned a value but never used. Allowed unused vars must match /^_/u  no-unused-vars

✖ 1 problem (1 error, 0 warnings)

** (exit) 1
    (mix 1.18.2) lib/mix/tasks/cmd.ex:74: Mix.Tasks.Cmd.run/1
    (mix 1.18.2) lib/mix/task.ex:495: anonymous fn/3 in Mix.Task.run_task/5
    (mix 1.18.2) lib/mix/task.ex:561: Mix.Task.run_alias/6
    (mix 1.18.2) lib/mix/cli.ex:107: Mix.CLI.run_task/2
    /home/runner/work/_temp/.setup-beam/elixir/bin/mix:2: (file)
Error: Process completed with exit code 1.
```

**Comment by Sancxo** [comment]
@bartblast Well this is weird because I didn't modified this file and even I can't find the mentioned variable `integer36` at line 41

Furthermore, `mix eslint` doesn't raise anything

**Comment by bartblast** [comment]
Weird indeed... Can you try merging the dev branch to see if that helps?

**Comment by Sancxo** [comment]
@bartblast Indeed, my fork was really out of sync. The problem is closed now 🙂

**Comment by bartblast** [comment]
@Sancxo Looks like there's a conflict on `call_graph.ex` now.

**Comment by bartblast** [comment]
@coderabbitai full review

File: assets/js/erlang/string.mjs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import Type from "../type.mjs";
+
+// IMPORTANT!
+// If the given ported Erlang function calls other Erlang functions, then list such dependencies in the "Deps" comment (see :erlang./=/2 for an example).
+// Also, in such case add respective call graph edges in Hologram.CallGraph.list_runtime_mfas/1.
+
+const Erlang_String = {
+  // Start replace/3
+  "replace/3": (string, pattern, replacement) => {
+    const replace = Erlang_String["replace/4"];
```
Comment (bartblast) [inline assets/js/erlang/string.mjs]: No need for a separate const IMO. It could be:
```js
return Erlang_String["replace/4"](string, pattern, replacement, Type.atom("leading"));
```

(that's the pattern used elsewhere)

File: assets/js/erlang/string.mjs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    if (!Type.isBinary(pattern)) {
+      Interpreter.raiseArgumentError(
+        Interpreter.buildArgumentErrorMsg(
+          1,
+          "not valid character data (an iodata term)",
+        ),
+      );
+    }
+
+    if (!Type.isBinary(replacement)) {
```
Comment (bartblast) [inline assets/js/erlang/string.mjs]: The current implementation only accepts binaries, but Erlang's `:string.replace` functions accept unicode:chardata() for both SearchPattern and Replacement parameters.

## What is chardata/iolist?

**Charlists** are lists of integer codepoints representing characters:
- Example: `[72, 101, 108, 108, 111]` represents `"Hello"`
- Shorthand: `~c"Hello"` (in Elixir) is the same as `[72, 101, 108, 108, 111]`
- Can contain any Unicode codepoint (0 to 0x10FFFF)

**Iolists (I/O lists)** are recursively nested lists that can contain:
- Integers (codepoints)
- Binaries (strings)
- Other iolists (arbitrarily nested)

**Examples of valid chardata:**

```elixir
# Flat charlist
[72, 101]              # "He"

# Nested iolist with binaries
[<<"He">>, "llo"]      # "Hello"

# Mixed nesting: integers, binaries, and nested lists
["H", [101, [<<"l">>]]]  # "Hel"

# Deeply nested (any nesting level is valid)
[[[72]], [[101]]]      # "He"
```

**How to convert chardata to Hologram boxed bitstring**

Good news: This is already implemented! Use `:unicode.characters_to_binary/1` which is already ported in `assets/js/erlang/unicode.mjs`.

**How to fix your implementation**

Replace lines 28-49 with:
```js
// Convert pattern from chardata to binary
const patternBinary = Erlang_Unicode["characters_to_binary/1"](pattern);

if (Type.isAtom(patternBinary) && patternBinary.value === "error") {
  Interpreter.raiseArgumentError(
    Interpreter.buildArgumentErrorMsg(
      1,
      "not valid character data (an iodata term)",
    ),
  );
}

// Convert replacement from chardata to binary  
const replacementBinary = Erlang_Unicode["characters_to_binary/1"](replacement);

if (Type.isAtom(replacementBinary) && replacementBinary.value === "error") {
  // Currently you throw HologramInterpreterError - this should be ArgumentError
  Interpreter.raiseArgumentError(
    Interpreter.buildArgumentErrorMsg(
      2,
      "not valid character data (an iodata term)",
    ),
  );
}

const stringText = Bitstring.toText(string);
const patternText = Bitstring.toText(patternBinary);
const replacementText = Bitstring.toText(replacementBinary);
```

Don't forget to:
1. Add dependency comment: `// Deps: [:unicode.characters_to_binary/1]`
2. Update call graph in `lib/hologram/compiler/call_graph.ex`:
`{{:string, :replace, 4}, {:unicode, :characters_to_binary, 1}},`
3. Import `Erlang_Unicode` at the top (check how other files do this)

**Test reference**

See how `:unicode.characters_to_binary/1` is tested in `test/javascript/erlang/unicode_test.mjs` for examples of handling charlists and nested iolists.

File: test/javascript/erlang/string_test.mjs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    it("returns a two-elements list with the first word at the begining and the tail at the end when the direction is set to :leading", () => {
+      const result = replace(
+        string_test,
+        Type.bitstring(" "),
+        Type.bitstring("."),
+        Type.atom("leading"),
+      );
+
+      assert.deepStrictEqual(result, Type.list(["Hello", ".", "World !"]));
```
Comment (bartblast) [inline test/javascript/erlang/string_test.mjs]: Actually there are 3 elements in the list.

File: test/javascript/erlang/string_test.mjs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    it("returns a two-elements list with the last word at the end and the rest at the begining when the direction is set to :trailing", () => {
+      const result = replace(
+        string_test,
+        Type.bitstring(" "),
+        Type.bitstring("."),
+        Type.atom("trailing"),
+      );
+
+      assert.deepStrictEqual(result, Type.list(["Hello World", ".", "!"]));
```
Comment (bartblast) [inline test/javascript/erlang/string_test.mjs]: Actually there are 3 elements in the list.

File: test/elixir/hologram/ex_js_consistency/erlang/string_test.exs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  Always update both together.
+  """
+
+  use Hologram.Test.BasicCase, async: true
+
+  @moduletag :consistency
+
+  describe "replace/3" do
+    test "returns a three-elements list with the first word at the begining, the replacement in the middle and the tail at the end when the direction is set to :leading" do
+      result = :string.replace("Hello World !", " ", "_", :leading)
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/string_test.exs]: You're actually testing `:string.replace/4` here.

File: test/elixir/hologram/ex_js_consistency/erlang/string_test.exs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  IMPORTANT!
+  Each Elixir consistency test has a related JavaScript test in test/javascript/erlang/rand_test.mjs
+  Always update both together.
+  """
+
+  use Hologram.Test.BasicCase, async: true
+
+  @moduletag :consistency
+
+  describe "replace/3" do
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/string_test.exs:78]: We just need to test the basic delegation (since `replace/3` delegates to `replace/4`) and have the tests for `replace/4`.

File: test/elixir/hologram/ex_js_consistency/erlang/string_test.exs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    test "returns a three-elements list with the first word at the begining, the replacement in the middle and the tail at the end when the direction is set to :leading" do
+      result = :string.replace("Hello World !", " ", "_", :leading)
+
+      assert result == ["Hello", "_", "World !"]
+    end
+  end
+
+  describe "replace/4" do
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/string_test.exs:86]: JS tests and Elixir tests don't match. The whole goal of the consistency tests is to verify we have the exact same behaviour for client-side and server-side. Hence the tests must be identical (and in identical order for easier maintenance).

File: test/javascript/erlang/string_test.mjs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+defineGlobalErlangAndElixirModules();
+
+// IMPORTANT!
+// Each JavaScript test has a related Elixir consistency test in test/elixir/hologram/ex_js_consistency/erlang/lists_test.exs
+// Always update both together.
+
+describe("Erlang_String", () => {
+  describe("replace/3", () => {
+    const replace = Erlang_String["replace/3"];
+    const string_test = Type.bitstring("Hello World !");
```
Comment (bartblast) [inline test/javascript/erlang/string_test.mjs]: string_test -> string?

File: test/elixir/hologram/ex_js_consistency/erlang/string_test.exs (hunk 6 above)
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/string_test.exs:78]: The current test suite covers basic functionality well, but could benefit from additional edge cases:

1. Pattern at the start of string: `replace("Hello", "He", "X", :leading)`
2. Pattern at the end of string: `replace("Hello", "lo", "X", :trailing)`
3. Consecutive patterns: `replace("aaa", "a", "X", :all)`
4. Unicode characters: r`eplace("Hello 👋 World", "👋", "🌍", :all)`

**Review by bartblast — CHANGES_REQUESTED** [review]
Besides my comments, could you also:
* pull the new changes from the dev branch
* address issues detected by CodeRabbit

File: assets/js/erlang/string.mjs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+    const stringText = Bitstring.toText(string);
+    const patternText = Bitstring.toText(pattern);
+    const replacementText = Bitstring.toText(replacement);
+
+    if (Bitstring.isEmpty(pattern) || !stringText.includes(patternText)) {
+      return Type.list([string]);
+    }
+
+    let splittedStringList, index;
```
Comment (bartblast) [inline assets/js/erlang/string.mjs:141]: You're shadowing the `index` parameter here.

**Comment by Sancxo** [comment]
Hi @bartblast, it should be good now :slightly_smiling_face:

**Comment by bartblast** [comment]
@Sancxo Could you pull the new changes from the dev branch?

File: test/javascript/erlang/string_test.mjs [hunk 10]
```diff
@@ -168,6 +169,183 @@ describe("Erlang_String", () => {
...
+            Type.atom("_"),
+            Type.bitstring(" "),
+            Type.atom("all"),
+          ),
+        "ArgumentError",
+        "errors were found at the given arguments:\n\n  * 1st argument: not valid character data (an iodata term)\n",
+      );
+    });
+
+    it("raises HologramInterpreterError if third argument is not a binary", () => {
```
Comment (bartblast) [inline test/javascript/erlang/string_test.mjs:240]: @Sancxo CI fails are caused by this test case, because the function returns boxed ArgumentError.

**Comment by bartblast** [comment]
Hey @Sancxo! Just a heads up - I'm planning to release all the ported functions in a few days. Any chance you'll be able to wrap this one up soon?

**Comment by Sancxo** [comment]
Hi @bartblast 

First, congrats for the Curiosum funding !
Then I think it's all good for this one :slightly_smiling_face:

**Comment by bartblast** [comment]
Hey @Sancxo, thanks, I appreciate it :) As for the PR - looks good - merging, thanks for your help! 💜
